Automattic/woocommerce-payments
# Add currency suffixes to key prices

Fixes #2098

#### Changes proposed in this Pull Request



This PR includes the backend price formatting application relative to the specifications defined in paJDYF-20W-p2. 

The fix includes changes both in PHP code and JS code. PHP code introduces two new classes: 

- **WC_Payments_Explicit_Price_Formatter** is responsible for adding filters to prices that need to show explicit prices

- **\WCPay\MultiCurrency\BackendCurrencies** is responsible for adding filters to prices *on the backend* to match the formats proposed in the new `locale-info` file. 

JS Part now uses `Intl\NumberFormat` library for the fallback currency formatting. The list on `@woocommerce\currency` package only contains 13 currency specs, and uses the fallback formatter for the other possible currencies. Now it uses the natively supported l10n library. The browser support of the library is 95.89% according to https://caniuse.com/mdn-javascript_builtins_intl_numberformat.

**Note:** The only part left from this fix is the Checkout Block total formatting, because its code is in WCAdmin and it uses `@woocommerce\price-format` which is also a separate library only included in that repo.

The fix includes these visual modifications:

<table>
<tr>
<th> Before </th><th> After </th>
</tr>

<tr><td colspan=2> Shop Page </td></tr>
<tr><td> <img src='https://user-images.githubusercontent.com/3295/127489233-f6c13fbe-61e4-4bfb-9c31-08c82e122c98.png'> </td><td> <img src='https://user-images.githubusercontent.com/3295/127474596-276b8d81-bb7b-4790-b692-982bbfb7f100.png'> </td></tr>

<tr><td colspan=2>Product Page</td></tr>
<tr><td> <img src='https://user-images.githubusercontent.com/3295/127489316-8aa46c40-0bb2-471b-a846-cf5e2fbe3dcd.png'> </td><td> <img src='https://user-images.githubusercontent.com/3295/127475000-33e50b81-d5e1-48cd-939c-5259c366d470.png'></td></tr>

<tr><td colspan=2>Cart Page</td></tr>
<tr><td> <img src='https://user-images.githubusercontent.com/3295/127489383-7d265409-028b-4e01-98c6-1fc280794eff.png'> </td><td> <img src='https://user-images.githubusercontent.com/3295/127475175-f03f07ef-ef82-42e2-b4f4-12dc91aca4c9.png'></td></tr>

<tr><td colspan=2>Checkout Page</td></tr>
<tr><td><img src='https://user-images.githubusercontent.com/3295/127505918-b96f05e3-5661-4eef-bf79-a570579b31bf.png'></td><td><img src='https://user-images.githubusercontent.com/3295/127505813-0930588a-349f-4382-aec2-02a59c60026b.png'></td></tr>

<tr><td colspan=2>Confirmation Page</td></tr>
<tr><td><img src='https://user-images.githubusercontent.com/3295/127489523-ec3f2be7-b54a-40e3-b8ba-963f4b201831.png'> </td><td> <img src='https://user-images.githubusercontent.com/3295/127482660-1f5c3685-23ac-40f3-aac0-c34819a01c52.png'></td></tr>

<tr><td colspan=2>Orders Metabox in WooCommerce Homescreen</td></tr>
<tr><td> <img src='https://user-images.githubusercontent.com/3295/127489056-f0a25f95-c477-4fab-9d0f-05dd80068395.png'> </td><td> <img src='https://user-images.githubusercontent.com/3295/127483591-ed70d771-b9b9-40a7-8dbc-0d2b36ddf89f.png'></td></tr>

<tr><td colspan=2>WooCommerce Orders Page</td></tr>
<tr><td> <img src='https://user-images.githubusercontent.com/3295/127488927-3d6a02b1-9e78-496f-86ef-4027f12d2589.png'> </td><td> <img src='https://user-images.githubusercontent.com/3295/127483825-f5c74027-8087-4280-8a6c-84f2f8a71102.png'></td></tr>

<tr><td colspan=2>Order Details Page</td></tr>
<tr><td> <img src='https://user-images.githubusercontent.com/3295/127488814-eff5a828-7d79-4ccd-bd33-becaa288b867.png'> </td><td> <img src='https://user-images.githubusercontent.com/3295/127485979-2822f66e-9b97-4629-ba5f-f4076b3d0f68.png'></td></tr>

<tr><td colspan=2>WCPay Overview Page</td></tr>
<tr><td> <img src='https://user-images.githubusercontent.com/3295/127488717-276de74b-2791-414e-8b6a-295045a3c6f4.png'> </td><td> <img src='https://user-images.githubusercontent.com/3295/127486056-994b8a72-05c8-44e0-a619-6aaf57803d49.png'></td></tr>

<tr><td colspan=2>Deposits Screen</td></tr>
<tr><td> <img src='https://user-images.githubusercontent.com/3295/127488631-bd242f44-e484-4bd2-b3cd-5acc950590d0.png'> </td><td> <img src='https://user-images.githubusercontent.com/3295/127486192-6b5d65c1-a19e-4ab5-935f-85f0e73c32dd.png'></td></tr>

<tr><td colspan=2>Transactions Screen</td></tr>
<tr><td> <img src='https://user-images.githubusercontent.com/3295/127488324-d5782fa9-20ec-4adf-8e02-7df640765821.png'> <img src='https://user-images.githubusercontent.com/3295/127488517-e7236ef0-1baa-40d5-8d75-713092af1058.png'> </td><td> <img src='https://user-images.githubusercontent.com/3295/127486284-e5dba2eb-6033-4157-8536-962972a68c67.png'><br><img src='https://user-images.githubusercontent.com/3295/127486391-e00d9d57-f716-4e4c-9aec-fd87f16c19ee.png'></td></tr>

<tr><td colspan=2>Transaction Details Page</td></tr>
<tr><td> <img src='https://user-images.githubusercontent.com/3295/127488443-76d990d1-4ea7-45d2-9671-b4ef6e994e5f.png'> </td><td> <img src='https://user-images.githubusercontent.com/3295/127486486-7e56b91f-66da-4170-a66b-2b7bdba8abc0.png'></td></tr>

<tr><td colspan=2>Disputes Screen</td></tr>
<tr><td> <img src='https://user-images.githubusercontent.com/3295/127488146-2d7f00a9-b169-4aa3-94ef-21a47a12b733.png'> </td><td> <img src='https://user-images.githubusercontent.com/3295/127486940-e0828e29-71b9-45bb-a70f-f49cebe1c6d5.png'></td></tr>

<tr><td colspan=2>Dispute Details Page</td></tr>
<tr><td> <img src='https://user-images.githubusercontent.com/3295/127487763-0c37a9ee-dcd7-4632-8ba9-66cc74dd6545.png'> </td><td> <img src='https://user-images.githubusercontent.com/3295/127487009-8011cf43-15ba-4547-aee6-0a38434c104e.png'></td></tr>

</table>








#### Testing instructions





* run `npm run test:js` to run Jest tests, updated existing ones with the new currency displays.
* run `npm run test:php` to run PHPUnit tests
* Visually check the specs from paJDYF-20W-p2 if they match the live output

-------------------
